### ISSUE TITLE
better performance out of the box

### ISSUE THREAD

**Issue by beckermr** [conda-forge/corrfunc-feedstock#2]
Hey @manodeep, I went ahead and packaged your awesome code here for ease of use. I've noticed, as you say very clearly, that with the preferred installation method (`pip`), the performance is a lot better. I assume this is because the compiled code is finding the vectorized instructions.

Would you be willing to help debug the builds here? I cannot figure out where the compiler settings are going wrong.

The build logs are [here](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=75721&view=logs). If you click into "Run docker build" and scroll down, you can find the actual logs from running the install to build the conda package.

**Comment by msarahan** [comment]
Compiler settings for GCC are from the compiler activation package. They default to targeting "nocona", which equates to sse3. We do this for compatibility. Many programs compiled to target newer CPUs will crash when run on old CPUs. In fact, when we first released our compilers in 2017, people complained that they did not have sse3.

Be careful when asking these flags. You should prefer a method of offering various optimization levels, either using multiple dispatch (fat binaries with many code paths for various instruction sets; mkl and openblas do this) or by creating multiple packages at different cpu support levels. Conda's new virtual package support could be used for the latter, but would take some development effort. If this package is already clever enough to detect CPU capabilities, you can hopefully just crank up the CPU timing to a much more modern one.

**Comment by beckermr** [comment]
I am pretty sure this package supports multiple dispatch: https://github.com/manodeep/Corrfunc/blob/master/theory/DD/countpairs_impl.c.src#L98

**Comment by manodeep** [comment]
Thanks @beckermr for setting up the conda-forge feedstock. I will take a look at the log and get back to you. In the meantime, here is my understanding of both conda builds and Corrfunc architecture. 

Yeah the primary difference with the `pip` installed version would be the compiler flag `-march=native`. Obviously, you can not add such a flag for an universal build, otherwise the package will crash with an illegal instruction error. 

While `Corrfunc` does support multiple dispatch, but that dispatch is only targeted for the different kernels. The remainder of the Corrfunc libraries are not hardware-aware, and therefore, if you compile the code targeting a later instruction set, then the build will likely crash again on older hardware. 

The correct way to do this would be the add `target` options to each of the kernels. For instance, the `AVX512F` kernel needs to be marked with `__attribute__ ((target ("avx512f")))`, and so on, with `default` for the `Fallback` kernel. Since this new attribute only kicks in when generating the universal builds, it might not even be necessary to extend such gcc-specific "target" options for the other comilers (`clang/icc`). 

Somewhat related, we seem to have uncovered a bug in `gcc7.3` [here](https://github.com/manodeep/Corrfunc/issues/193). @beckermr If you are building with these target keywords attached to the kernels, then just use a more recent gcc. If you don't have access to a more recent gcc, then build with `icc`, adjusting the function attributes to the `icc` guidelines.

**Comment by beckermr** [comment]
I'm not following all of this, but note that the conda builds for osx use clang. Thus enabling this for both linux and osx would probably require at least supporting both gcc and clang. It looks like [clang supports](https://clang.llvm.org/docs/AttributeReference.html#target) GNU-style attributes with target. However, IDK if actual options passed are the same. I'm pretty out of my depth here!

**Comment by manodeep** [comment]
@beckermr The changes are required for every kernel within every pair-counter. Taking the example of `theory/DD`, here are the changes that would be necessary:

1. [This line](https://github.com/manodeep/Corrfunc/blob/master/theory/DD/countpairs_kernels.c.src#L33) would need `__attribute__ ((target ("avx512f")))` after the function parameters and the closing parenthesis

2. [This line](https://github.com/manodeep/Corrfunc/blob/master/theory/DD/countpairs_kernels.c.src#L293) would need `__attribute__ ((target ("avx")))` after the function parameters

3. [This line](https://github.com/manodeep/Corrfunc/blob/master/theory/DD/countpairs_kernels.c.src#L607) would need `__attribute__ ((target ("sse4_2")))` 

The `Fallback` kernels do not need any separate annotations. The `wp` kernel also has a separate `avx2` kernel which would need `__attribute__ ((target ("avx2")))`

Once this is done, the runtime dispatch would pick out the correct kernel based on the runtime CPU characteristics and all should be well. 

Hope this makes sense. Happy to clarify, if not.

**Comment by manodeep** [comment]
@beckermr Here's a bit of code that might be helpful (probably in `utils/macros.h`)

```C
#if !defined(_MSC_VER)
#define CORRFUNC_TARGET(isa) __attribute__ ((target (isa))) 
#else 
/* MS: Unverified but should work on Windows */
#define CORRFUNC_TARGET(isa) __declspec__ ((target (isa))) 
#endif
```

Now, all you need to do is append `CORRFUNC_TARGET("avx512f")`, `CORRFUNC_TARGET("avx")`, `CORRFUNC_TARGET("sse4.2")` etc in front of every kernel. For example:

```C
static inline int countpairs_avx512_intrinsics_DOUBLE(const int64_t N0, DOUBLE *x0, ...
/* will become */

CORRFUNC_TARGET("avx512f") 
static inline int countpairs_avx512_intrinsics_DOUBLE(const int64_t N0, DOUBLE *x0, ...
```

I got this specific implementation from [libalgebra.h](https://github.com/mklarqvist/libalgebra/blob/master/libalgebra.h#L286-L290). To see what the target code looks like in production, [see here](https://github.com/mklarqvist/libalgebra/blob/master/libalgebra.h#L786)

If you can confirm that this approach works on a combination of compilers and runtime configurations, then I am happy to update the Corrfunc repo. PR's are of course welcome!

**Comment by manodeep** [comment]
Responding here about the `archspec` comment - if I understood correctly, that would create multiple Corrfunc builds targeting different cpu architectures. If so, then seems like a reasonable way to go forward, without any source code modifications.

Thanks again for maintaining this conda recipe. 

That reminds me, do you mind creatiing a PR to the Corrfunc repo updating the README to include this `conda install -c conda-forge corrfunc` method of installing?

**Comment by beckermr** [comment]
We need to patch out any native flags in the corrfunc source.

**Comment by beckermr** [comment]
@lgarrison for viz

**Comment by lgarrison** [comment]
I did some experiments with `__attribute__((target))`, but strangely I think it might only work in C++, not C.  That's what the [documentation online](https://gcc.gnu.org/onlinedocs/gcc/Function-Multiversioning.html) seems to say, and my tests confirmed, but maybe I'm not looking in the right place, or there's another syntax for C...

So we might be back to the version where we do one compilation unit per kernel, or one library per instruction set.  Or maybe the conda "virtual packages" mentioned above would be a faster route to happiness?

**Comment by beckermr** [comment]
The virtual packages thing is not yet merged. I think for now to fix the conda-forge build, we need to remove those flags with a patch.

**Comment by lgarrison** [comment]
I agree, that sounds fine.  I guess there's not much point in `-march=nocona` either, because our SSE kernels require SSE 4.2, and nocona stops at SSE 3!

**Comment by manodeep** [comment]
Is there an option to build with `icc` and the `-axHost` flag within conda-build? 

gcc >= 6 supports multi-versioning in C but the syntax I quoted was wrong - should be something like `    __attribute__((target_clones("sse4.2",  "avx", "avx2", "avx512f", "default")))`. See [here](https://lwn.net/Articles/691932/)

**Comment by beckermr** [comment]
No, we cannot build with `icc` in conda-forge. We actually have access to the compilers but we can't use them in the CI services due to various restrictions on our access.

**Comment by beckermr** [comment]
A fat binary like openblas is the best solution now if we want to ship higher levels of optimization.

**Comment by manodeep** [comment]
What's the compiler in conda-forge? As in, is it possible to add those lines to *all* the functions in the `gridlink*` files, and recompile with function multi-versioning?

**Comment by beckermr** [comment]
See the docs here. https://conda-forge.org/docs/maintainer/infrastructure.html#compilers-and-runtimes

We use a mix of gcc and clang. It should be possible. Openblas does it.

**Comment by lgarrison** [comment]
Protecting every function individually with `target_clones` is ugly, but I agree it should work (but see below).  And in that case, I guess one would use `-march=icelake-server` or something very modern, as long as the conda-forge compilers support it. The conda-forge CPUs shouldn't have to.

Do we have any functions that we rely on being inlined?  It looks like `target_clones` will not produce multiple versions for functions that are called from the annotated functions, unless they are inlined.  And adding `target_clones` to the inlined function results in a version check for every call: https://hannes.hauswedell.net/post/2017/12/09/fmv/.  So we'll need to be careful there.

**Comment by manodeep** [comment]
@lgarrison Because the `gridlink*` functions are invoked twice at most, **and** are usually not the bottleneck in computing, it should be okay regardless of any inlining . 

@beckermr Do you have a path forward to create such a Corrfunc binary with function multi-versioning?

**Comment by beckermr** [comment]
I don't have the relevant expertise to make the build work like this.

**Comment by manodeep** [comment]
Brainstorming here for a solution. Here are some additional lines of code for `common.mk` to be inserted [here](https://github.com/manodeep/Corrfunc/blob/master/common.mk#L526):

```makefile
GRIDLINK_CFLAGS:=$(CFLAGS)
ifneq ($(CONDA_BUILD),)
  $(warning Building with conda-build, disabling arch-specific codes for gridlink files)
  $(warning Before removing "-march=native", GRIDLINK_CFLAGS = "[$(GRIDLINK_CFLAGS)]")
  GRIDLINK_CFLAGS:=$(filter-out -march=native,$(GRIDLINK_CFLAGS))
  $(warning After removing "-march=native", GRIDLINK_CFLAGS = "[$(GRIDLINK_CFLAGS)]")
endif
```

And these lines go into `rules.mk` inserted [here](https://github.com/manodeep/Corrfunc/blob/master/rules.mk#L59)

```makefile
gridlink_%_double.o: gridlink_%_double.c gridlink_%_double.h
	$(CC) -DDOUBLE_PREC $(GRIDLINK_CFLAGS) $(INCLUDE) $(EXTRA_INCL) -c $< -o $@

gridlink_%_float.o: gridlink_%_float.c gridlink_%_float.h
	$(CC) -DNDOUBLE_PREC $(GRIDLINK_CFLAGS) $(INCLUDE) $(EXTRA_INCL) -c $< -o $@
```

Can you check whether this patch works out?

**Comment by beckermr** [comment]
This patch causes a core dump w/ an illegal instruction.

**Comment by lgarrison** [comment]
I am not sure if this is the immediate issue, but one concern that comes to mind with removing `-march=native` for gridlink but keeping it for the impl drivers is that not everything in the impl files is protected by multiple-dispatch.  E.g. any of the code in [`countpairs_DOUBLE()`](https://github.com/manodeep/Corrfunc/blob/master/theory/DD/countpairs_impl.c.src#L136) is allowed to emit SIMD instructions that may cause an illegal instruction.  This is the function that invokes multiple-dispatch to call the correct SIMD kernel safely, but it itself is not protected this way.  We would need to put the kernels in a separate compilation unit and remove `-march=native` from the impl driver compilation to be strictly correct.  Right now the kernels are just `#include`-d into the impl: https://github.com/manodeep/Corrfunc/blob/master/theory/DD/countpairs_impl.c.src#L17

In practice, I don't know if this is causing the immediate error, because there isn't too much math happening in `countpairs_DOUBLE()`.  But there is some, like mins, maxes, thread-reduce, and rupp^2.

**Comment by manodeep** [comment]
@lgarrison Yup compiling and linking the kernels separately with the `-march=native` is precisely the correct solution but then the kernels might no longer be inlined. More importantly, that will require a bit of tinkering with the Makefiles to make sure that the dependencies are correctly noted. Such a change would be significant enough that we would have to implement it within the Corrfunc repo itself. 

Another option might be to add the `-march=skylake-avx512` in kernels files (with gcc pragmas with gcc >= 6) and then remove that `-march=native` at the end of the kernel files. Not sure if this is possible to implement though - I know gcc diagnostic options can be added/removed via source code but unsure about code-gen options.
